rise/static/main.js

```diff
@@ -16,79 +16,143 @@ define([
+    // use the ones in livereveal/legacy first
+    // so they get overridden if redefined in rise
+    let config_section_legacy = new configmod.ConfigSection(
       'livereveal',
       {base_url: utils.get_body_data("baseUrl")});
-    config_section.load();
-
-    /* dummy empty config section to load
-     * the metadata + default as a ConfigWithDefaults object
-     */
-    var _config_section = new configmod.ConfigSection(
-      '_livereveal',
+    // trigger an asynchronous load
+    config_section_legacy.load();
+    let config_section = new configmod.ConfigSection(
+      'rise',
       {base_url: utils.get_body_data("baseUrl")});
-    _config_section.load();
-
-    var final_config;
-
-    var rise_meta = Jupyter.notebook.metadata.livereveal;
-
-    if(rise_meta !== undefined && Object.keys(rise_meta).length > 0){
-      final_config = $.extend(true, default_config, rise_meta);
-      final_config = new configmod.ConfigWithDefaults(_config_section, final_config);
-      console.log("RISE metadata detected. Using ONLY RISE metadata on top of the default config. Custom config disabled.")
-    } else {
-      final_config = new configmod.ConfigWithDefaults(config_section, default_config);
-      console.log("No (or empty) RISE metadata. Using ONLY custom config (if exist) on top of the default config.")
-    }
-
-    return final_config
+    config_section.load();
 
+    // this is also a ConfigSection object as per notebook/static/services/config.js 
+    let nbext_configurator = Jupyter.notebook.config;
+    nbext_configurator.load();
+
+    // with Promise.all we can wait for all 3 configs to have loaded
+    return Promise.all([
+      config_section_legacy.loaded,
+      config_section.loaded,
+      nbext_configurator.loaded,
+    ]).then(
+        // and now we can compute the layered config
+        function() {
+          // 1) initialize with hardwired defaults
+          $.extend(true, complete_config, hardwired_config);
+          // 2a) and 2b)
+          $.extend(true, complete_config, config_section_legacy.data);
+          $.extend(true, complete_config, config_section.data);
+          // 3)
+          $.extend(true, complete_config, nbext_configurator.data.rise);
+          // 4a) from the notebook metadata
+          let metadata_legacy = Jupyter.notebook.metadata.livereveal;
+          $.extend(true, complete_config, metadata_legacy);
+          // 4b) ditto
+          let metadata = Jupyter.notebook.metadata.rise;
+          $.extend(true, complete_config, metadata);
+          // console.log("complete_config is OK");
+        });
+  }
+    
+  /*
+   * this function is a heuristic that says if this notebook seems to
+   * be meant to be a slideshow.
+   * this primarily is for autolaunch, so that somebody who would
+   * enable autolaunch in her ~/.jupyter/ area would not
+   * see RISE trigger on every single notebook
+   *
+   * xxx note that this might take too long on large notebooks
```

Comment:
That's a good point, maybe we can break the loop when a slide (or subslide) is found? That should reduce the time we spend on large notebooks IF they have a slide (subslide) cell.
In any case, we can do this later, no need to change for now.

Reply:
right; actually when a slide or subslide is found, then the function returns immediately

I was more concerned about people who would have RISE installed but who would work on large non-slides notebooks

but indeed this must not keep us from moving forward, it's an easy fix that can be made later on